### DISCUSSION TITLE
[Bug]: Black Screens

### DISCUSSION THREAD

**Discussion by StShAzzy** [project64/project64#2268] [General]
### Expected behavior?

![image](https://user-images.githubusercontent.com/90798128/162241976-6c6ff3f5-ad7c-460f-a805-b3c8153b37d7.png)
everything working fine

### Actual behavior?

![image](https://user-images.githubusercontent.com/90798128/162241795-0b8618ea-098b-4d09-8489-cca121f9bd21.png)
sound keeps playing but the screen is just dead

### Steps to reproduce the problem

not sure how but, change the GlideN64 settings while playing a game the screen should go black but the game keep running
atleast that's what i do to cause the screen to go black, or entering fullscreen and exiting it.
loading a save state sometimes fixes the screen going black... sometimes
also there's a chance the black screen persists even through resets my only option in this case is restarting the whole program, that means i have to wait for the donation screen again, which is a inconvenience
 

### Error messages?

not a single one :/

### What game were you playing?

Title: Super Mario 64

Nintendo 64 game or Nintendo 64 Disk Drive: N64

ROM information: CRC1 0x635A2BFF CRC2 0x8B022326 CIC Chip CIC-NUS-6102 MD5 49FBF9C744C458405F685D42B762A0BE

Save type: (Attach game save or save state here)

Original or modified ROM?: Original

### Specifications and build details

Project64: 3.0.1.5664.2df3434 (include commit version number and whether you are using the
32-bit or 64-bit version)

OS: Windows 10 Pro 64bits version 1903 build: 18362.1500

CPU: Intel Pentium G2030 @ 3.0GHz

RAM: 4 GB

Graphics: https://www.techpowerup.com/gpu-specs/hd-graphics.c1250

### Using included plugins

- [X] I can confirm I am using the included RSP, audio, input, and graphics plugins

### Plugins used while the issue occurred

Graphics plugin (and HLE or LLE):
Audio plugin:
Input plugin:
RSP plugin:
![image](https://user-images.githubusercontent.com/90798128/162237090-44434a72-1837-4b1d-9eb1-51f14de52635.png)

### Copyright Compliance

- [X] There are no infringing files attached to this bug report.

**Comment by dftf-stu** [comment]
@StShAzzy: given your CPU launched in the [middle of 2013](https://ark.intel.com/content/www/us/en/ark/products/74749/intel-pentium-processor-g2030-3m-cache-3-00-ghz.html), you might be better using the "Project64 Video" plug-in, rather than the "GLideN64".

I've been advised here before that older Intel GPUs and/or their drivers tend to be buggy in how they implement newer versions of OpenGL: if you use "Project64 Video" it won't attempt to use anything-past OpenGL 1.1, whereas OpenGL 3.3 is what "GLideN64" will look to use.  So limiting yourself down-to OpenGL 1.1 via the older plug-in might help.

You could also try downloading a ZIP of the latest "Dev" release, 4.0.0.5871, from [here](https://www.pj64-emu.com/nightly-builds) and check if the issue is still present.

**Comment by Squall-Leonhart** [comment]
Thats not correct, Glide64 based Project64 Video uses OpenGL 2+, 1.1 has no capability of implementing the Glide api's used.

**Comment by dftf-stu** [comment]
@Squall-Leonhart: why does it say OpenGL 1.1 [here](https://github.com/project64/project64#readme) then, on the landling-page for this project...?

![Screenshot 2022-08-05 174700](https://user-images.githubusercontent.com/83965343/183123652-7ac93e90-35d8-4f86-bcf8-df90cafb9d4a.png)

Surely it should say "OpenGL 2.0 capable (Project64 Video)" then?
Bit confusing to us users if the official info we're being given is wrong!

Would it take someone long to update the README.md?

**Comment by Squall-Leonhart** [comment]
Thats because it assumes the max capability of the voodoo5 that glide64 was originally made for, but ignores all the additional capabilities added to the custom wrapper that relies on 2.x

**Comment by dftf-stu** [comment]
@Squall-Leonhart: so if it "ignores all the additional capabilities added", then the information is out-of-date then?

So my point stands: users can only go by what information is presented to them.  And I read it as: "If I have a computer that is running at-least Windows XP, with a GPU that is capable of at-least OpenGL 1.1 then using the "Project64 Video" plug-in, I should expect that the latest stable-version should work, and be capable of playing any ROM listed in green".

Whereas actually the case is "The system-specs there are based on the really old versions of _Project64_, from the late 90s and mid-2000s, back from the original maintainer, going-by the plug-ins as they were then, not how they are now.  We know that the latest versions of our app don't run on XP (unless you edit the config-file to turn-off the Discord support) or Vista (due to an API change), but also won't advise that.  Oh, and if a ROM is listed as "compatible" and in green in the ROM Browser you should NOT take that to mean it actually is compatible "out-of-the-box", it just represents the possible maximum level of compatibility you can achieve with the right CPU, GPU, GPU driver, plug-ins, emulator settings, ROM-specific settings and enhancements... though as we don't have any sort of "game database" to advise what settings you'll need to use, your only options are to use _Google_ and hope you can find a post (such-as on _Reddit_) where someone has also had issues with the same game, or either post on here and ask for help, or ask in our _Discord_."

Are you surprised users get confused? :p

**Comment by StShAzzy** [comment]
my GPU is capable of running up to OpenGL 4.0, OpenGL 3.3 shouldn't be causing issues, i'll try the build out, even through i've already figured out what was causing the black screen, it was the N64 style depth comparison

**Comment by dftf-stu** [comment]
@StShAzzy: just because on-paper, in the technical specs, it says "OpenGL 4.0", doesn't mean that either the GPU itself or the drivers that _Windows_ uses to talk to the GPU, actually support it fully.  Look at things like USB 3.0 or HDMI ports: they all look the same physically, but between devices the speeds, wattage, screen-resolutions, refresh-rates, and so-on, they offer varies.

And as _Project64_ doesn't offer a "software-renderer" (where the entire N64 game runs only within your CPU), that means they are at the mercy of bad drivers or GPUs not actually fully-supporting specs they say they do.

(Hopefully, in future, a software-renderer could be considered, as I see users have asked about a macOS port of _Project64_ in the past -- and OpenGL has been depreciated by Apple since 2018's macOS 10.14 "Mojave", meaning it is no-longer active supported and bugs are unlikely to get fixed.  Or that OpenGL could get removed entirely in a future release!)

**Comment by StShAzzy** [comment]
i also tested and the bug persists, thanks for helping even if i already figured out the problem myself, what about DirectX and Vulkan Support?

**Comment by Squall-Leonhart** [comment]
The vendors diverge where their implementation of the extensions are concerned, nvidia is guilty of going abit and beyond which those in the mesa community cry about as "violations", AMD and Intel are closer to the actual spec but even then each have had bugs in their implementations that only get picked up once developers begin using them.

Intel is also notorious for exposing capabilities that the hardware doesn't actually properly support in the way the driver implements the extension.

**Comment by dftf-stu** [comment]
_Project64_ does come with a plug-in called "Jabo's Direct3D8" (for _DirectX_), but this was last-updated (according to the copyright) in 2008, and others on here have advised me will be removed in future major versions of the app.

I've not heard anything about support for _Vulkan_ being added by-default in-future, but you could search for a plug-in called "ParaLLEl-RDP", which apparently will unofficially give you that support.  This (https://github.com/mudlord/pj64-parallelrdp) was apparently the link to where you downloaded the files for _Project64,_ but it appears broken now.

_Vulkan_ isn't officially supported on macOS; look into _MoltenVK_ as an open-source workaround.

**Comment by flagrama** [comment]
https://github.com/Rosalie241/parallel-rsp
https://github.com/Rosalie241/pj64-parallelrdp

**Comment by Squall-Leonhart** [comment]
there are still specification differences for vulkan among the 3 vendors that can break things.

Your IGP doesn't support vulkan and only has FL11 compliance for D3D.

**Comment by dftf-stu** [comment]
@StShAzzy: looks like flagrama has offered a working link, so go [here](https://github.com/Rosalie241/pj64-parallelrdp/releases), download the "pj64-parallelrdp.dll" file, and then while _Project64_ is closed (not running), go into the folder where the app itself is located, then go into the "Plugin" folder, then "GFX" and copy it inside there.

The next time you start _Project64_, go to Options > Configuration > Plugins and it should be listed in the top dropdown list.

**Reply by flagrama** [reply]
The RSP is required as well. The RSP and RDP are one package so need to work properly in tandem.
